dbms/src/Common/getNumberOfPhysicalCPUCores.cpp

```diff
@@ -15,51 +15,67 @@
 #include <Common/Exception.h>
 #include <Common/getNumberOfPhysicalCPUCores.h>
 
-#include <thread>
-
-#if defined(__x86_64__)
-#include <cpuid/libcpuid.h>
+#if defined(OS_LINUX)
+#include <cmath>
+#include <fstream>
 #endif
 
-namespace DB
+#include <thread>
+
+#if defined(OS_LINUX)
+static int32_t readFrom(const char * filename, int default_value)
 {
-namespace ErrorCodes
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+        return default_value;
+    int idata;
+    if (infile >> idata)
+        return idata;
+    else
+        return default_value;
+}
+
+/// Try to look at cgroups limit if it is available.
+static uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
 {
-extern const int CPUID_ERROR;
+    uint32_t quota_count = default_cpu_count;
+    /// Return the number of milliseconds per period process is guaranteed to run.
+    /// -1 for no quota
+    int cgroup_quota = readFrom("/sys/fs/cgroup/cpu/cpu.cfs_quota_us", -1);
+    int cgroup_period = readFrom("/sys/fs/cgroup/cpu/cpu.cfs_period_us", -1);
+    if (cgroup_quota > -1 && cgroup_period > 0)
+        quota_count = static_cast<uint32_t>(ceil(static_cast<float>(cgroup_quota) / static_cast<float>(cgroup_period)));
+
+    return std::min(default_cpu_count, quota_count);
 }
-} // namespace DB
+#endif
 
-unsigned getNumberOfPhysicalCPUCores()
+static unsigned getNumberOfPhysicalCPUCoresImpl()
 {
-    unsigned res = 0;
-#if defined(__x86_64__)
+    unsigned cpu_count = std::thread::hardware_concurrency();
+
+    /// Most of x86_64 CPUs have 2-way Hyper-Threading
+    /// Aarch64 and RISC-V don't have SMT so far.
+    /// POWER has SMT and it can be multiple way (like 8-way), but we don't know how ClickHouse really behaves, so use all of them.
 
-    cpu_raw_data_t raw_data;
-    cpu_id_t data;
-    if (0 == cpuid_get_raw_data(&raw_data) && 0 == cpu_identify(&raw_data, &data) && data.num_logical_cpus != 0)
-    {
-        res = data.num_cores * data.total_logical_cpus / data.num_logical_cpus;
-    }
+#if defined(__x86_64__)
+    /// Let's limit ourself to the number of physical cores.
+    /// But if the number of logical cores is small - maybe it is a small machine
+    /// or very limited cloud instance and it is reasonable to use all the cores.
+    if (cpu_count >= 32)
```

Comment:
A little weired here, this cpu_count = 32, its max stream will be 16, smaller than cpu_count = 24. That's machine with more cpus has smaller max_streams.